File: ruby/lib/static_array.rb
# This class emulates a static array which has a fixed size and location in memory

class StaticArray

  attr_reader :length

  def initialize(length = 8)
    @store = Array.new(length)
    @length = length
  end

  # fixed place in memory means time O(1) for accessing and setting
  def [](index)
    raise "index out of bounds" if index < 0 || index >= @length || !index.is_a?(Integer)
    @store[index]
  end

  def []=(index, value)
    raise "index out of bounds" if index < 0 || index >= @length || !index.is_a?(Integer)
    @store[index] = value
  end

end

File: js/src/StaticArray.js

const StaticArray = function (length = 8) {
  this.store = new Array(length);
  this.length = length;
};

StaticArray.prototype.get = function (index) {
  if (index >= this.length || index < 0) {
    throw "index out of bounds";
  }
  return this.store[index];
};

StaticArray.prototype.set = function (index, value) {
  if (index >= this.length || index < 0 || typeof index !== 'number') {
    throw "index out of bounds";
  }
  this.store[index] = value;
  return value;
};


module.exports = StaticArray;
